tools/tpm2_import.c

```diff
@@ -566,3 +597,138 @@ int tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
+    bool result = true;
+    switch (alg) {
+    case TPM2_ALG_AES :
+        obj->algorithm = TPM2_ALG_AES;
+        obj->keyBits.aes = 128;
+        obj->mode.aes = TPM2_ALG_CFB;
+        break;
+    case TPM2_ALG_NULL :
+        obj->algorithm = TPM2_ALG_NULL;
+        break;
+    default:
+        LOG_ERR("The algorithm type input(0x%x) is not supported!", alg);
+        result = false;
+        break;
+    }
+    return result;
+}
+
+static int tpm_import(ESYS_CONTEXT *ectx) {
+    bool result;
+    TPM2B_DATA enc_key;
+    TPM2B_PUBLIC public = TPM2B_EMPTY_INIT;
+    TPM2B_PRIVATE duplicate;
+    TPM2B_ENCRYPTED_SECRET encrypted_seed;
+    TPM2B_PRIVATE *imported_private;
+    tpm2_loaded_object parent_object_context;
+    TPMT_SYM_DEF_OBJECT sym_alg;
+
+    result = tpm2_util_object_load(ectx, ctx.parent_ctx_arg,
+                                &parent_object_context);
+    if (!result) {
+      return 1;
+    }
+
+    if (ctx.parent_auth_str) {
+        result = tpm2_auth_util_from_optarg(ectx, ctx.parent_auth_str,
+            &ctx.auth.session_data, &ctx.auth.session);
+        if (!result) {
+            LOG_ERR("Invalid parent key authorization, got\"%s\"", ctx.parent_auth_str);
+            return 1;
+        }
+    }
+
+    result = set_key_algorithm(ctx.input_enc_key_type, &sym_alg);
+    if(!result) {
+        return 1;
+    }
+
+    /* Symmetric key */
+    if(ctx.input_enc_key_file) {
+        enc_key.size = 16;
+        result = files_load_bytes_from_path(ctx.input_enc_key_file, enc_key.buffer, &enc_key.size);
+        if(!result) {
+            LOG_ERR("Failed to load symmetric encryption key\"%s\"", ctx.input_enc_key_file);
+            return 1;
+        }
+        if(enc_key.size != 16) {
+            LOG_ERR("Invalid AES key size, got %u bytes, expected 16", enc_key.size);
+            return 1;
+        }
+    }
+
+    /* Private key */
+    result = files_load_private(ctx.input_duplicate_file, &duplicate);
+    if(!result) {
+        LOG_ERR("Failed to load duplicate \"%s\"", ctx.input_duplicate_file);
+        return 1;
+    }
+
+    /* Encrypted seed */
+    result = files_load_encrypted_seed(ctx.input_seed_file, &encrypted_seed);
+    if(!result) {
+        LOG_ERR("Failed to load encrypted seed \"%s\"", ctx.input_seed_file);
+        return 1;
+    }
+
+    /* Public key */
+    result = files_load_public(ctx.public_key_file, &public);
+    if(!result) {
+        LOG_ERR(":( Failed to load public key \"%s\"", ctx.public_key_file);
+        return 1;
+    }
+
+    public.publicArea.authPolicy.size = sizeof(public.publicArea.authPolicy.buffer);
+    result = files_load_bytes_from_path(ctx.policy,
+                public.publicArea.authPolicy.buffer, &public.publicArea.authPolicy.size);
+    if (!result) {
+        return 1;
+    }
+
+    result = do_import(
+            ectx,
+            parent_object_context.tr_handle,
+            &encrypted_seed, 
+            ctx.input_enc_key_file == NULL ? NULL : &enc_key,
```

Comment:
Just initialize enc_key size to 0, then just check in the routine if size is 0, nothing to do versus if NULL nothing to do.

Reply:
My understanding here is that if symmetricAlg is null then encryptionKey should be NULL too. I think then I would have to put the check 'enc_key.size == 0 ? NULL : &enc_key' inside do_import(). 

Is that correct?

Reply:
yep